content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+With the information above, we can formulate two problems that we have to solve in order to get to more responsive user interfaces:
+
+1. **Long-running tasks cause frame drops.** We need to make sure all of our tasks are small and can be completed within a couple of milliseconds so that we can run them within one frame.
+
+2. **Different tasks have different priorities.** In the example application above, we saw that prioritizing the user input leads to a better experience overall. To do this, we need a way to define the order and to schedule tasks accordingly.
+
+## Concurrent React and the Scheduler
+
+_⚠️ Warning: The following APIs are not yet stable and will change. I will do my best to keep this post updated._
+
+To implement a properly scheduled user interface with React, we have to look into two new React features:
+
+- **Concurrent React (also known as Time Slicing).** With the help of the new [Fiber architecture](https://www.youtube.com/watch?v=ZCuYPiUIONs) rewrite that was released with React 16, React can now pause during rendering and yield[^2] to the main thread.
+
+  ➡️ With this feature, we’re able to split long-running rendering tasks into small chunks.
+
+- **Scheduler.** A general purpose cooperative main thread scheduler developed by the React Core team that makes it possible to register callbacks with different priority levels in the browser.
+
+  At the time of writing this article, the priority levels are:
+
+  - `Immediate` for tasks that need to run synchronously.
+  - `UserBlocking` (250ms timeout) for tasks that should run as the result of a user interaction (e.g. a button click).
+  - `Normal` (5s timeout) for rendering non-interactive parts of the screen.
+  - `Low` (10s timeout) for tasks that can be deferred but must still complete eventually (e.g. an analytics notification).
+  - `Idle` (no timeout) for tasks that do not have to run at all. (e.g. hidden offscreen content).
+
+  The timeouts for each priority level are necessary to make sure that lower priority work still runs even if we have so much higher priority work to do that the higher priority work could run continuously. In scheduling algorithms, this problem is referred to as [starvation](<https://en.wikipedia.org/wiki/Starvation_(computer_science)>). The timeouts give us the guarantee that every scheduled task will eventually run. For example, we won’t miss a single analytics notification even if we have ongoing animations in our app.
+
+  ➡️ With this feature, we can schedule tasks with different priorities.
+
+## Scheduling in Action
+
+Let’s see how we can use these features to make an app feel a lot more responsive. To do this, we’ll take a look at [ScheduleTron 3000](https://github.com/philipp-spiess/scheduletron3000), an app I built that allows us to highlight a search term in a list of names. Let’s take a look at the implementation first:
+
+<!-- prettier-ignore -->
+```js
+// The app shows a search box and a list of names. The list is
```

Comment:
Which or that? Only on searchValue state variable possible, or could there be multiple?

"// The app shows a search box and a list of names. The list is
// controlled by the searchValue state variable which is updated
// by the search box."

Reply:
Only one possible so I think it's "which".